### PR TITLE
[NFC] Centralize TypeChecker Flags With TypeCheckerOptions

### PR DESCRIPTION
**Pull request by CodaFi**
Merge TypeCheckingFlags, TypeChecker's Options and the TypeChecker-Oriented FrontendOptions into (the one true heir) `TypeCheckerOptions`.

Use this to simplify a bunch of things that only needed type checkers just for this bit of state and, importantly, pull all of this out of the TypeChecker.

@davidungar Could you make sure I haven't done something terrible to frontend argument parsing?
@harlanhaskins Same, but for function body skipping?
@rintaro For the ParserUnit parts (particularly the `llvm::SaveAndRestore`s to emulate the old behavior of passing `None` to the options).

### CODE REVIEW

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please test

**Comment by harlanhaskins** [comment]
Function body skipping stuff looks god, but why do all those tests now need `import Foundation`?

**Comment by CodaFi** [comment]
For some reason, we're actually emitting the diagnostic for these now.  I suspect it's because something was calling `performTypeChecking` without passing the right arguments through.

**Comment by theblixguy** [comment]
Could we pass `-disable-objc-attr-requires-foundation-module` in the tests?

**Comment by CodaFi** [comment]
@theblixguy Good point.  I'll push this through once the tests pass

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please test

**Comment by brentdax** [comment]
Some other things we might want to move over from LangOptions:

* DebugConstraintSolver
* DebugConstraintSolverAttempt
* DebugConstraintSolverOnLines
* DebugForbidTypecheckPrefix
* SolverMemoryThreshold
* SolverBindingThreshold
* SolverShrinkUnsolvedThreshold
* SolverDisableShrink
* DisableConstraintSolverPerformanceHacks
* SolverEnableOperatorDesignatedTypes

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

**Review by harlanhaskins — APPROVED** [review]

**Review by rintaro — APPROVED** [review]
Parser related changes LGTM.

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

File: tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp [hunk 1]
```diff
@@ -1240,7 +1241,7 @@ class SwiftDocumentStructureWalker: public ide::SyntaxModelWalker {
     }
 
     Consumer.beginDocumentSubStructure(StartOffset, EndOffset - StartOffset,
-                                       Kind, AccessLevel, SetterAccessLevel,
```
Comment (CodaFi) [inline tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp]: Bad merge, whoops.

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/llvm-project#240

@swift-ci please smoke test

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**